pimd/pim_register.c

```diff
@@ -131,32 +156,58 @@ int pim_register_stop_recv(struct interface *ifp, uint8_t *buf, int buf_size)
 	pim_parse_addr_ucast(&source, buf, buf_size);
 	sg.src = source.u.prefix4;
 
+	rp = RP(pim_ifp->pim, sg.grp);
+
+	if (sg.src.s_addr == rp->rpf_addr.u.prefix4.s_addr)
+		sg.src.s_addr = INADDR_ANY;
+
 	upstream = pim_upstream_find(pim, &sg);
-	if (!upstream) {
+
+	if ((!upstream) && (sg.src.s_addr != INADDR_ANY))
 		return 0;
-	}
 
-	if (PIM_DEBUG_PIM_REG)
-		zlog_debug("Received Register stop for %s", upstream->sg_str);
+	/*
+	 * RFC 7761 Sec 4.4.1
+	 * Handling Register-Stop(*,G) Messages at the DR:
+	 *   A Register-Stop(*,G) should be treated as a Register-Stop(S,G) for
+	 * all (S,G) Register state machines that are not in the NoInfo state.
+	 */
+
+	/* No starg upstream. loop upstream list to get all sg */
+	if ((!upstream) && (sg.src.s_addr == INADDR_ANY)) {
+		struct pim_upstream *up;
+
+		frr_each (rb_pim_upstream, &pim->upstream_head, up) {
+			if (sg.grp.s_addr != up->sg.grp.s_addr)
+				continue;
+
+			if (PIM_DEBUG_PIM_REG)
+				zlog_debug("Executing Reg stop for %s",
+					   up->sg_str);
+
+			pim_reg_stop_upstream(pim, up);
+		}
 
-	switch (upstream->reg_state) {
-	case PIM_REG_NOINFO:
-	case PIM_REG_PRUNE:
 		return 0;
-	case PIM_REG_JOIN:
-		upstream->reg_state = PIM_REG_PRUNE;
-		pim_channel_del_oif(upstream->channel_oil, pim->regiface,
-				    PIM_OIF_FLAG_PROTO_PIM, __func__);
-		pim_upstream_start_register_stop_timer(upstream, 0);
-		pim_vxlan_update_sg_reg_state(pim, upstream,
-			false/*reg_join*/);
-		break;
-	case PIM_REG_JOIN_PENDING:
-		upstream->reg_state = PIM_REG_PRUNE;
-		pim_upstream_start_register_stop_timer(upstream, 0);
+	}
+
+	/* Starg upstream present. Loop through children */
+	if (upstream && (sg.src.s_addr == INADDR_ANY)) {
+		struct pim_upstream *child;
+		struct listnode *up_node;
+
+		for (ALL_LIST_ELEMENTS_RO(upstream->sources, up_node, child)) {
+			if (PIM_DEBUG_PIM_REG)
+				zlog_debug("Executing Reg stop for %s",
+					   child->sg_str);
+
+			pim_reg_stop_upstream(pim, child);
+		}
+
 		return 0;
 	}
 
+	pim_reg_stop_upstream(pim, upstream);
```

Comment:
how will this not crash?  We've already established upstream can be null here.

Reply:
If upstream was null, we would have returned already. following conditions return.
if ((!upstream) && (sg.src.s_addr != INADDR_ANY))
if ((!upstream) && (sg.src.s_addr == INADDR_ANY))

Reply:
1) Please do not mark my review comments as resolved that is for me to do.

2) Please relook and notice there is no return case for the !upstream case. 

How did you actually test this?  I would like to see some additional test cases in topotest that show this new functionality working correctly.

Reply:
Line numbers 167 and 191 has return for 166 and 177 respectively. Sorry if I'm missing something here.

Reply:
We have a loop for !upstream and a *,G received and then we exit that loop come down here and immediately crash

Reply:
@donaldsharp there is a return 0 after the loop you mentioned.
        if ((!upstream) && (sg.src.s_addr == INADDR_ANY)) {
                struct pim_upstream *up;

                frr_each (rb_pim_upstream, &pim->upstream_head, up) {
                        if (sg.grp.s_addr != up->sg.grp.s_addr)
                                continue;

                        if (PIM_DEBUG_PIM_REG)
                                zlog_debug("Executing Reg stop for %s",
                                           up->sg_str);

                        pim_reg_stop_upstream(pim, up);
                }

                return 0; <====================== after the loop returns in this statement.
        }

Regarding the testing, this test is being done by ANVL conformance suite. This particular case needs a custom packet to be sent out.
We will be upstreaming wide range of topotests covering scenarios other than conformance(50+).

Reply:
@sarav511 omitting irrelevant code here, what you have is:
```
if (upstream && ...) {
    ...
}

pim_reg_stop_upstream(pim, upstream);
```
The first thing `pim_reg_stop_upstream` does is dereference `upstream` with no nullity check. So either you know `upstream` will never be null, in which case you need to remove the `if (upstream ...)`, or it can sometimes be null, in which case you need to check before calling `pim_reg_stop_upstream`.

Reply:
@qlyoung Good point. will remove the useless upstream there in if. When control reaches there, definitely we have non-null upstream.

Reply:
@qlyoung Done

Reply:
Still looks wrong to me, can you elaborate?

Reply:
I dont see a way upstream will be null while executing that last instruction..

if ((!upstream) && (sg.src.s_addr != INADDR_ANY)) {
---
---
return 0;
}

if ((!upstream) && (sg.src.s_addr == INADDR_ANY)) {
--
--
return 0;
}

After this point in code upstream cannot be NULL. The above two conditions cover when upstream is NULL and they return from within 'if' block without falling through. Then we have only remaining two non-null upstream conditions.

if (sg.src.s_addr == INADDR_ANY) { <== this is equalent to  if (upstream && sg.src.s_addr == INADDR_ANY)
---
---
return 0;
}

pim_reg_stop_upstream(pim, upstream); <======= the only possible condition it executes this instruction is when (upstream && sg.src.s_addr != INADDR_ANY)

@qlyoung  Not sure what is wrong here. Can you please help me understand what I'm missing here?

Reply:
@qlyoung We have a scope of reducing one more condition here. Let me change and push it in. Hope it will clarify more.

Reply:
@qlyoung I have simplified a bit. With the recent push, you can notice that beyond line 197(return on !upstream), upstream definitely exists.